### PR TITLE
C++: Don't strip specifiers in `Node.getType`

### PR DESCRIPTION
**Pull request by MathiasVP**
We were calling `getUnspecifiedType` instead of `getUnderlyingType` in a bunch of places when computing the type of a dataflow node, and since `getUnspecifiedType` strips away specifiers such as `const` and `volatile` we could get into a situation where a node's `getType` returned `int` while the same node's `asExpr().getType()` returned `const int`.

The fix is to use `getUnderlyingType` instead of `getUnspecifiedType` since `getUnderlyingType` only resolves `typedefs` without stripping specifiers.

Finally, I also fixed a bug where some nodes were not correctly specified as being dataflow nodes wrapping a glvalue instruction or operand. This also fixed some inconsistency errors.

There still are some type-related bugs, but they seem a bit more difficult to fix. I'll write up an issue for the remaining ones.

Commit-by-commit review recommended.

### CODE REVIEW

**Comment by MathiasVP** [comment]
DCA reports one additional new result that looks like:
```cpp
const char* patterns[] = { "%s", "file=%s", "\"%s\"", "file=\"%s\"" };
char buf[1 * K];
for (size_t i = 0; i < ARRAY_SIZE(patterns); i++) {
  int ret = jio_snprintf(buf, sizeof(buf), patterns[i], TestLogFileName);
  /* ... */
}
```
where we claim that `patterns[i]` isn't a constant format string. I think that's a query FP that I expect to be fixed by https://github.com/github/codeql/pull/15516. So I don't think that should block this PR.

File: cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll [hunk 1]
```diff
@@ -942,11 +942,14 @@ private Type getTypeImpl0(Type t, int indirectionIndex) {
   or
   indirectionIndex > 0 and
   exists(Type stripped |
-    stripped = stripPointer(t.stripTopLevelSpecifiers()) and
-    // We need to avoid the case where `stripPointer(t) = t` (which can happen on
-    // iterators that specify a `value_type` that is the iterator itself). Such a type
-    // would create an infinite loop otherwise. For these cases we simply don't produce
-    // a result for `getTypeImpl`.
+    stripped = stripPointer(t) and
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll:945]: What's the reason behind this change?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/ir/dataflow/internal/DataFlowUtil.qll]: Fixed in dd3d70134cbd069c9fa9daa7474d55d359daa487

**Review by jketema — APPROVED** [review]
